C/services/notification-plugin-interfaces/python/python_plugin_interface.cpp

```diff
@@ -0,0 +1,657 @@
+#define SHIM_SCRIPT_NAME "notification_shim"
+
+using namespace std;
+
+extern "C" {
+
+extern PLUGIN_INFORMATION *Py2C_PluginInfo(PyObject *);
+extern void logErrorMessage();
+extern PLUGIN_INFORMATION *plugin_info_fn();
+extern PLUGIN_HANDLE plugin_init_fn(ConfigCategory *);
+extern void plugin_shutdown_fn(PLUGIN_HANDLE);
+
+// Reconfigure entry point for rule and delivery plugings
+void notification_plugin_reconfigure_fn(PLUGIN_HANDLE,
+					const std::string&);
+// Notification rule plugin entry points
+std::string plugin_triggers_fn(PLUGIN_HANDLE handle);
+bool plugin_eval_fn(PLUGIN_HANDLE handle,
+		    const std::string& assetValues);
+std::string plugin_reason_fn(PLUGIN_HANDLE handle);
+//Notificztion deelivery plugin entry point
+bool plugin_deliver_fn(PLUGIN_HANDLE handle,
+			const std::string& deliveryName,
+			const std::string& notificationName,
+			const std::string& triggerReason,
+			const std::string& customMessage);
+
+// Python object to instantiate
+PyObject* pModule = NULL;
+
+// PluginName
+string sPluginName;
+
+/**
+ * Constructor for PythonPluginHandle
+ *    - Load python 3.5 interpreter
+ *    - Set sys.path and sys.argv
+ *    - Import shim layer script and pass plugin name in argv[1]
+ *    - Set plygin_type (notificationRule or notificationDelivery in rgv[2]
+ *
+ * @param    pluginName		The plugin name to load
+ * @param    pluginPathName	The plugin pathname
+ */
+void *PluginInterfaceInit(const char *pluginName, const char * pluginPathName)
+{
+	// Set plugin name
+	sPluginName = pluginName;
+	// Get FOGLAMP_ROOT dir
+	string foglampRootDir(getenv("FOGLAMP_ROOT"));
+
+	string path = foglampRootDir + SHIM_SCRIPT_REL_PATH;
+	string name(SHIM_SCRIPT_NAME);
+
+	// Python 3.5  script name
+	std::size_t found = path.find_last_of("/");
+	string pythonScript = path.substr(found + 1);
+	string shimLayerPath = path.substr(0, found);
+
+	// Embedded Python 3.5 program name
+	wchar_t *programName = Py_DecodeLocale(name.c_str(), NULL);
+	Py_SetProgramName(programName);
+	PyMem_RawFree(programName);
+
+	string foglampPythonDir = foglampRootDir + "/python";
+
+	// Embedded Python 3.5 initialisation
+	Py_Initialize();
+	PyEval_InitThreads();
+	PyThreadState* save = PyEval_SaveThread(); // release Python GIT
+	PyGILState_STATE state = PyGILState_Ensure();
+
+	Logger::getLogger()->info("NotificationPlugin PythonInterface %s:%d: "
+				  "shimLayerPath=%s, shimFile=%s, "
+				  "foglampPythonDir=%s, plugin '%s",
+				   __FUNCTION__,
+				   __LINE__,
+				   shimLayerPath.c_str(),
+				   name.c_str(),
+				   foglampPythonDir.c_str(),
+				   sPluginName.c_str());
+
+	// Set Python path for embedded Python 3.5
+	// Get current sys.path - borrowed reference
+	PyObject* sysPath = PySys_GetObject((char *)"path");
+	PyList_Append(sysPath, PyUnicode_FromString((char *) shimLayerPath.c_str()));
+	PyList_Append(sysPath, PyUnicode_FromString((char *) foglampPythonDir.c_str()));
```

Comment:
It might be worth checking the existence of the shim layer path stuff - just to be able to give a user friendly error back rather than just failing.

Reply:
yes, next commits